File: apps/android/app/src/main/java/com/sanvya/app/ui/notifications/PushController.kt
package com.sanvya.app.ui.notifications

import android.Manifest
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import com.google.firebase.messaging.FirebaseMessaging
import com.sanvya.app.R
import com.sanvya.app.data.repository.nowIso
import com.sanvya.app.domain.repository.PushRepository
import kotlinx.coroutines.suspendCancellableCoroutine
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * Everything the app does with the OS notification system, in one place.
 *
 * Ported from `apps/web/src/notifications/push.ts`, whose four exported
 * functions this mirrors: read the permission, subscribe, send a local test,
 * and unsubscribe. Mirrors iOS's PushController.swift.
 *
 * **Permission is NOT asked at launch any more.** `MainActivity.onCreate` used
 * to call `askNotificationPermission()` in its first frame, before the user had
 * seen a single screen — the classic way to earn a permanent refusal from
 * someone who had no idea what they were being asked about. Web asks only when
 * the toggle is turned on; so does this now.
 */
class PushController(private val context: Context) : KoinComponent {
    private val pushRepository: PushRepository by inject()

    /** "granted" | "denied" | "notDetermined", as `pushState()` expects. */
    fun permission(): String {
        if (!supported()) return "denied"
        // Below API 33 there is no runtime permission at all -- notifications
        // are granted at install time, and the only way to refuse is the
        // system settings screen, which `areNotificationsEnabled()` reflects.
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
            return if (NotificationManagerCompat.from(context).areNotificationsEnabled()) "granted" else "denied"
        }
        val granted = ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) ==
            PackageManager.PERMISSION_GRANTED
        if (granted) return "granted"
        // Android cannot distinguish "never asked" from "refused twice" without
        // remembering that we asked. `asked` is that memory: without it every
        // refusal would read as notDetermined and the row would offer a switch
        // that can no longer do anything.
        return if (asked()) "denied" else "notDetermined"
    }

    fun supported(): Boolean = true

    /**
     * The current device token, or null when Firebase cannot mint one (no Play
     * Services, no network on first run, a stale `google-services.json`).
     *
     * Wrapped rather than callback-style so the caller can `try`/`catch` it
     * like any other failure -- the old code used `addOnCompleteListener` and
     * dropped a failed task on the floor with `return@addOnCompleteListener`.
     */
    suspend fun currentToken(): String = suspendCancellableCoroutine { cont ->
        FirebaseMessaging.getInstance().token
            .addOnSuccessListener { token -> if (cont.isActive) cont.resume(token) }
            .addOnFailureListener { e -> if (cont.isActive) cont.resumeWithException(e) }
    }

    /**
     * Register this device against [userId].
     *
     * Assumes permission is already granted -- the permission request itself is
     * an Activity concern (it needs a result launcher), so the screen asks and
     * then calls this.
     */
    suspend fun register(userId: String): String {
        val token = currentToken()
        pushRepository.registerToken(token, PLATFORM, userId, nowIso())
        return token
    }

    /** Drop this device's token. The OS permission is deliberately left alone. */
    suspend fun unregister() {
        val token = runCatching { currentToken() }.getOrNull() ?: return
        pushRepository.unregisterToken(token)
    }

    /**
     * Fire a notification locally — no server, no FCM.
     *
     * Web's `sendTestNotification` explains why this earns its place: it proves
     * permission and the delivery path work on THIS device, so when it shows
     * and real alerts do not, the gap is the server dispatch and not the phone.
     * That is the single most useful thing a support conversation can
     * establish.
     */
    fun sendTest(): Boolean {
        if (permission() != "granted") return false
        ensureChannel()
        val notification = NotificationCompat.Builder(context, CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_notification)
            .setContentTitle(TEST_TITLE)
            .setContentText(TEST_BODY)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .setAutoCancel(true)
            .build()
        return try {
            NotificationManagerCompat.from(context).notify(TEST_ID, notification)
            true
        } catch (e: SecurityException) {
            // Permission revoked between the check and the post. Rare, but the
            // API is documented to throw and a crash here would be absurd.
            false
        }
    }

    /**
     * The channel every Sanvya notification arrives on.
     *
     * Created on demand rather than at launch: a channel appears in the user's
     * system settings the moment it exists, and listing notification categories
     * for an app that has never asked to notify anyone is noise.
     */
    fun ensureChannel() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return
        val manager = context.getSystemService(NotificationManager::class.java) ?: return
        if (manager.getNotificationChannel(CHANNEL_ID) != null) return
        manager.createNotificationChannel(
            NotificationChannel(CHANNEL_ID, CHANNEL_NAME, NotificationManager.IMPORTANCE_DEFAULT).apply {
                description = CHANNEL_DESCRIPTION
            },
        )
    }

    /** Remembers that we have shown the runtime prompt at least once. */
    fun markAsked() {
        prefs().edit().putBoolean(ASKED_KEY, true).apply()
    }

    private fun asked(): Boolean = prefs().getBoolean(ASKED_KEY, false)

    private fun prefs() = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    companion object {
        const val PLATFORM = "android"
        const val CHANNEL_ID = "sanvya_alerts"

        private const val CHANNEL_NAME = "Alerts"
        private const val CHANNEL_DESCRIPTION =
            "Bills, budgets, low balances and unusual spend."
        private const val PREFS_NAME = "sanvya_push"
        private const val ASKED_KEY = "postNotificationsAsked"
        private const val TEST_ID = 4242

        // English on all three platforms, because it is English on web --
        // `sendTestNotification` writes these as literals too. See
        // PARITY_AUDIT's i18n row.
        const val TEST_TITLE = "Sanvya"
        const val TEST_BODY = "Test notification — you're all set to receive alerts."
    }
}

File: apps/ios/App/ViewModels/PushController.swift
import Foundation
import UIKit
import UserNotifications
import Factory
import Domain
import Data

/// Everything the app does with the OS notification system, in one place.
///
/// Ported from `apps/web/src/notifications/push.ts`, whose four exported
/// functions this mirrors: read the permission, subscribe, send a local test,
/// and unsubscribe. Mirrors Android's PushController.kt.
///
/// **Permission is NOT asked at launch any more.** `AppDelegate` used to call
/// `requestAuthorization` in `didFinishLaunchingWithOptions` — the system alert
/// arriving before the user had seen a single screen. On iOS that is worse than
/// impolite: the prompt can be shown exactly once, a "Don't Allow" is
/// permanent, and asking without context is on App Review's own list of
/// rejection reasons. Web asks only when the toggle is turned on; so does this
/// now.
@MainActor
public final class PushController {
    @Injected(\.pushRepository) private var pushRepository

    public init() {}

    /// The APNs token from the most recent successful registration.
    ///
    /// A `static` on purpose: `didRegisterForRemoteNotificationsWithDeviceToken`
    /// is an `AppDelegate` callback and there is no way to hand it an instance.
    /// Written there, read here.
    public private(set) nonisolated(unsafe) static var deviceToken: String?

    static func setDeviceToken(_ token: String?) { deviceToken = token }

    /// "granted" | "denied" | "notDetermined", as `pushState()` expects.
    ///
    /// `.provisional` counts as granted: it IS a working delivery channel (a
    /// quiet one), and reporting it as "not asked" would put the app back into
    /// prompting a user who is already receiving alerts.
    public func permission() async -> String {
        let settings = await UNUserNotificationCenter.current().notificationSettings()
        switch settings.authorizationStatus {
        case .authorized, .provisional, .ephemeral: return "granted"
        case .denied: return "denied"
        case .notDetermined: return "notDetermined"
        @unknown default: return "notDetermined"
        }
    }

    public let supported = true

    /// Ask for permission, and on a yes, start APNs registration.
    ///
    /// Returns what the user chose. The TOKEN does not arrive here — it comes
    /// back asynchronously through the app delegate — which is why `register`
    /// below waits for it rather than assuming one exists.
    public func requestPermission() async -> Bool {
        let granted = (try? await UNUserNotificationCenter.current()
            .requestAuthorization(options: [.alert, .sound, .badge])) ?? false
        if granted { UIApplication.shared.registerForRemoteNotifications() }
        return granted
    }

    /// Register this device against `userId`.
    ///
    /// Assumes permission is already granted.
    @discardableResult
    public func register(userId: String) async throws -> String {
        UIApplication.shared.registerForRemoteNotifications()
        guard let token = await waitForToken() else {
            throw PushError(message: "Couldn't get a notification token from Apple. Check the network and try again.")
        }
        try await pushRepository.registerToken(
            token: token, platform: Self.platform, userId: userId, lastSeenIso: nowIso()
        )
        return token
    }

    /// Drop this device's token. The OS permission is deliberately left alone.
    public func unregister() async throws {
        guard let token = Self.deviceToken else { return }
        try await pushRepository.unregisterToken(token: token)
    }

    /// APNs answers on its own schedule, and on a cold launch the token is not
    /// there yet. Poll briefly rather than fail instantly — a hard failure here
    /// would make "turn push on" flaky on exactly the launch where the user
    /// first tries it.
    private func waitForToken() async -> String? {
        for _ in 0..<Self.tokenWaitAttempts {
            if let token = Self.deviceToken { return token }
            try? await Task.sleep(for: .milliseconds(Self.tokenWaitStepMs))
        }
        return Self.deviceToken
    }

    /// Fire a notification locally — no server, no APNs.
    ///
    /// Web's `sendTestNotification` explains why this earns its place: it proves
    /// permission and the delivery path work on THIS device, so when it shows
    /// and real alerts do not, the gap is the server dispatch and not the phone.
    /// That is the single most useful thing a support conversation can
    /// establish.
    public func sendTest() async -> Bool {
        guard await permission() == "granted" else { return false }
        let content = UNMutableNotificationContent()
        content.title = Self.testTitle
        content.body = Self.testBody
        // A 1-second trigger, not nil. A nil trigger fires immediately, and a
        // notification that fires while its own app is foregrounded is
        // suppressed unless the delegate says otherwise -- so the user taps
        // "send test" and sees nothing at all, which is the exact opposite of
        // what the button is for.
        let request = UNNotificationRequest(
            identifier: Self.testIdentifier,
            content: content,
            trigger: UNTimeIntervalNotificationTrigger(timeInterval: 1, repeats: false)
        )
        do {
            try await UNUserNotificationCenter.current().add(request)
            return true
        } catch {
            return false
        }
    }

    public struct PushError: Error, LocalizedError {
        public let message: String
        public var errorDescription: String? { message }
    }

    static let platform = "ios"
    private static let testIdentifier = "sanvya-test"
    private static let tokenWaitAttempts = 20
    private static let tokenWaitStepMs = 150

    // English on all three platforms, because it is English on web —
    // `sendTestNotification` writes these as literals too. See PARITY_AUDIT's
    // i18n row.
    static let testTitle = "Sanvya"
    static let testBody = "Test notification — you're all set to receive alerts."
}
